lib/IRGen/IRGenModule.cpp

```diff
@@ -1185,63 +1175,200 @@ static bool isFirstObjectFileInModule(IRGenModule &IGM) {
   return containingModule->getFiles().front() == file;
 }
 
-void IRGenModule::emitAutolinkInfo() {
-  // Collect the linker options already in the module (from ClangCodeGen).
+static bool
+doesTargetAutolinkUsingAutolinkExtract(const SwiftTargetInfo &TargetInfo,
+                                       const llvm::Triple &Triple) {
+  if (TargetInfo.OutputObjectFormat == llvm::Triple::ELF && !Triple.isPS4())
+    return true;
+
+  if (TargetInfo.OutputObjectFormat == llvm::Triple::Wasm)
+    return true;
+
+  if (Triple.isOSCygMing())
+    return true;
+
+  return false;
+}
+
+namespace {
+
+struct AutolinkKind {
+  enum ValueTy {
+    LLVMLinkerOptions,
+    LLVMDependentLibraries,
+    SwiftAutoLinkExtract,
+  };
+
+  ValueTy Value;
+
+  AutolinkKind(ValueTy value) : Value(value) {}
+  AutolinkKind(const AutolinkKind &kind) : Value(kind.Value) {}
+
+  StringRef getSectionNameMetadata();
+
+  void collectEntriesFromLibraries(SmallVectorImpl<llvm::MDNode *> &Entries,
+                                   const ArrayRef<LinkLibrary> &AutolinkEntries,
+                                   IRGenModule &IGM);
+
+  void writeEntries(const ArrayRef<llvm::MDNode *> &Entries,
+                    llvm::NamedMDNode *Metadata, IRGenModule &IGM);
+
+  static AutolinkKind create(const SwiftTargetInfo &TargetInfo,
+                             llvm::Triple Triple, IRGenLLVMLTOKind LLVMLTOKind);
+};
+} // namespace
+
+StringRef AutolinkKind::getSectionNameMetadata() {
   // FIXME: This constant should be vended by LLVM somewhere.
-  auto *Metadata = Module.getOrInsertNamedMetadata("llvm.linker.options");
-  for (llvm::MDNode *LinkOption : Metadata->operands())
-    AutolinkEntries.push_back(LinkOption);
+  switch (Value) {
+  case AutolinkKind::LLVMDependentLibraries:
+    return "llvm.dependent-libraries";
+  case AutolinkKind::LLVMLinkerOptions:
+  case AutolinkKind::SwiftAutoLinkExtract:
+    return "llvm.linker.options";
+  }
 
-  // Remove duplicates.
-  llvm::SmallPtrSet<llvm::MDNode *, 4> knownAutolinkEntries;
-  AutolinkEntries.erase(std::remove_if(AutolinkEntries.begin(),
-                                       AutolinkEntries.end(),
-                                       [&](llvm::MDNode *entry) -> bool {
-                                         return !knownAutolinkEntries.insert(
-                                                   entry).second;
-                                       }),
-                        AutolinkEntries.end());
-
-  const bool AutolinkExtractRequired =
-      (TargetInfo.OutputObjectFormat == llvm::Triple::ELF && !Triple.isPS4()) ||
-      TargetInfo.OutputObjectFormat == llvm::Triple::Wasm ||
-      Triple.isOSCygMing();
-
-  if (!AutolinkExtractRequired) {
+  llvm_unreachable("Unhandled AutolinkKind in switch.");
+}
+
+void AutolinkKind::collectEntriesFromLibraries(
+    SmallVectorImpl<llvm::MDNode *> &Entries,
+    const ArrayRef<LinkLibrary> &AutolinkEntries, IRGenModule &IGM) {
```

Comment:
ArrayRef is a value type that refers to an unowned const array. You don't need const or to pass by reference.